Add support for custom keyboard shortcuts

joajfreitas opened the pull request.

joajfreitas committed:
Implement support for custom keyboard shortcuts

tiborschneider commented on src/lib.rs:
```diff
@@ -125,9 +125,10 @@ impl<T> RofiChild<T> {
         let status = self.p.wait()?;
-        if status.success() {
+        let code = status.code().unwrap();
+        if status.success() || (code >= 10 && code <= 30)  {
```
Could you explain why you check the exit code to be between 10 and 30?

joajfreitas replied:
Sure! When you set and use a custom keyboard shortcut in rofi the return code will correspond to the used shortcut. This information will not be provided in the stdout.

The limits I've set here are not exactly correct has you can see here under ROFI_RETV: https://davatorium.github.io/rofi/1.7.5/rofi-script.5/

I'll clean this up and ping you soon

tiborschneider commented on src/lib.rs:
```diff
@@ -259,6 +262,14 @@ where
         self
     }
 
+    ///
```
Can you provide helpful comments here that describe how to use that feature? (Or at least a link to the rofi documentation where people can lookup the syntax)?

tiborschneider reviewed (commented):
Thanks for your contribution! Can you explain why you need the exact exit code, and why you treat exit code 10 to 30 as a success?

joajfreitas commented:
@tiborschneider I've removed the change to the `run` and `run_index` return types. Now rofi's exit code can be fetched with a getter function.

Can you take a look at the PR?

tiborschneider commented on src/lib.rs:
```diff
@@ -259,6 +296,14 @@ where
         self
     }
 
+    /// Set the message of the rofi window (-mesg). Only available in dmenu mode.
+    /// Docs: <https://github.com/davatorium/rofi/blob/next/doc/rofi-dmenu.5.markdown#dmenu-specific-commandline-flags>
+    pub fn message(&mut self, message: impl Into<String>) -> &mut Self {
+        self.args.push("-mesg".to_string());
+        self.args.push(message.into());
+        self
+    }
+
```
Didn't know that was possible, thanks!

tiborschneider commented on src/lib.rs:
```diff
@@ -435,6 +492,9 @@ pub enum Error {
     ConfigErrorMessageAndOptions,
+    /// User used a custom keyboard shortcut
+    #[error("User used a custom keyboard shortcut")]
+    CustomKeyboardShortcut,
```
Do you think we can just attach the code specific exit code here? That would make more sense to me, because it would pop up directly where one matches on the result.

tiborschneider commented on src/lib.rs:
```diff
@@ -474,12 +534,20 @@ mod rofitest {
+        let mut r = Rofi::new(&options);
+        match r.message("Press Alt+n").kb_custom(1, "Alt+n").run() {
+            Err(Error::CustomKeyboardShortcut) => {
+                assert_eq!(r.exit_code(), Some(10))
```
having the result directly in `Error::CustomKeyboardShortcut` would allow you to directly get the shortcut ID, without the need to subtract 9 (as it seems from your test case)

tiborschneider commented on src/lib.rs:
```diff
@@ -445,7 +505,7 @@ mod rofitest {
         let empty_options: Vec<String> = Vec::new();
         match Rofi::new(&options).prompt("choose c").run() {
-            Ok(ret) => assert!(ret == "c"),
+            Ok(ret) => assert!(ret == "c".to_string()),
```
Why is that one necessary now?

tiborschneider commented on src/lib.rs:
```diff
@@ -277,6 +322,13 @@ where
         self
     }
 
+    ///
```
Could you please put a comment here? Explain what `id` is, and what syntax to use with `shortcut`. Also, provide a link to the docs.

tiborschneider reviewed (commented):
Thanks for the revision. See my comments.

joajfreitas committed:
Address review points

joajfreitas commented:
@tiborschneider I've addressed your concerns. Rofi doesn't seem to have great documentation for the -kb-custom flags so I added a link straight to the source code.

Hope it looks good now!

tiborschneider commented:
Thank you so much for your contribution! It is now published under version 0.4
